File: lib/premium-analysis/savings-estimator.ts
import type { AnomalyFlag, SavingsEstimate } from '@/lib/premium-analysis/agents-types'
import { yearMonthKeyFromEngineMonthLabel } from '@/lib/premium-analysis/analysis-summary'

const RECOVERY_RATE: Partial<Record<AnomalyFlag['type'], { low: number; high: number }>> = {
  fuel_over_eia: { low: 0.5, high: 1 },
  contract_discount_shortfall: { low: 0.5, high: 1 },
  address_correction: { low: 0.8, high: 1 },
  large_package: { low: 0.25, high: 0.75 },
  additional_handling: { low: 0.25, high: 0.5 },
  avoidable_expedited: { low: 0.3, high: 0.7 },
  declared_value: { low: 0.5, high: 1 },
  accessorial_rate_high: { low: 0.1, high: 0.25 },
  weight_gap_high: { low: 0.15, high: 0.35 },
  monthly_spend_spike: { low: 0, high: 0 },
}

function monthsInDataset(monthlySpend: Array<{ month: string }>): number {
  const keys = new Set<string>()
  for (const m of monthlySpend) {
    const k = yearMonthKeyFromEngineMonthLabel(m.month)
    if (k) keys.add(k)
  }
  return Math.max(1, keys.size)
}

/** Scale overlapping flag totals so summed period amounts do not exceed total spend. */
export function capFlagAmountsBySpend(
  flags: AnomalyFlag[],
  periodTotalSpend: number
): Map<string, number> {
  const byType = new Map<string, number>()
  for (const flag of flags) {
    byType.set(flag.type, (byType.get(flag.type) ?? 0) + flag.amount)
  }
  const rawSum = [...byType.values()].reduce((s, a) => s + a, 0)
  if (periodTotalSpend <= 0 || rawSum <= periodTotalSpend) return byType
  const scale = periodTotalSpend / rawSum
  return new Map([...byType.entries()].map(([t, a]) => [t, a * scale]))
}

export function estimateSavings(
  anomalyFlags: AnomalyFlag[],
  monthlySpend: Array<{ month: string; totalCost?: number }>,
  periodTotalSpend?: number
): SavingsEstimate {
  const months = monthsInDataset(monthlySpend)
  const periodSpend =
    periodTotalSpend ?? monthlySpend.reduce((s, m) => s + (m.totalCost ?? 0), 0)
  const byType = capFlagAmountsBySpend(anomalyFlags, periodSpend)

  let opportunities = [...byType.entries()].map(([type, periodAmount]) => {
    const rates = RECOVERY_RATE[type as AnomalyFlag['type']] ?? { low: 0.2, high: 0.5 }
    const annualized = (periodAmount / months) * 12
    return {
      type,
      periodAmount,
      annualizedLow: annualized * rates.low,
      annualizedHigh: annualized * rates.high,
    }
  })

  let low = opportunities.reduce((s, o) => s + o.annualizedLow, 0)
  let high = opportunities.reduce((s, o) => s + o.annualizedHigh, 0)

  const annualizedSpendCap = periodSpend > 0 ? (periodSpend / months) * 12 : 0

  if (annualizedSpendCap > 0 && high > annualizedSpendCap) {
    const scale = annualizedSpendCap / high
    opportunities = opportunities.map((o) => ({
      ...o,
      annualizedLow: o.annualizedLow * scale,
      annualizedHigh: o.annualizedHigh * scale,
    }))
    low *= scale
    high = annualizedSpendCap
  }

  return {
    low: +low.toFixed(2),
    high: +high.toFixed(2),
    annualizedBasisMonths: months,
    opportunities: opportunities.sort((a, b) => b.annualizedHigh - a.annualizedHigh),
  }
}

File: scripts/premium_analysis_mirror/savings_estimator.py
"""Savings estimator — mirrors lib/premium-analysis/savings-estimator.ts."""

from __future__ import annotations

from typing import Any

from .primitives import year_month_key_from_engine_month_label

RECOVERY_RATE: dict[str, dict[str, float]] = {
    "fuel_over_eia": {"low": 0.5, "high": 1.0},
    "contract_discount_shortfall": {"low": 0.5, "high": 1.0},
    "address_correction": {"low": 0.8, "high": 1.0},
    "large_package": {"low": 0.25, "high": 0.75},
    "additional_handling": {"low": 0.25, "high": 0.5},
    "avoidable_expedited": {"low": 0.3, "high": 0.7},
    "declared_value": {"low": 0.5, "high": 1.0},
    "accessorial_rate_high": {"low": 0.1, "high": 0.25},
    "weight_gap_high": {"low": 0.15, "high": 0.35},
    "monthly_spend_spike": {"low": 0.0, "high": 0.0},
}


def _months_in_dataset(monthly_spend: list[dict[str, Any]]) -> int:
    keys = set()
    for m in monthly_spend:
        k = year_month_key_from_engine_month_label(m.get("month", ""))
        if k:
            keys.add(k)
    return max(1, len(keys))


def cap_flag_amounts_by_spend(flags: list[dict[str, Any]], period_total_spend: float) -> dict[str, float]:
    by_type: dict[str, float] = {}
    for flag in flags:
        t = flag["type"]
        by_type[t] = by_type.get(t, 0.0) + float(flag.get("amount") or 0)

    raw_sum = sum(by_type.values())
    if period_total_spend <= 0 or raw_sum <= period_total_spend:
        return by_type

    scale = period_total_spend / raw_sum
    return {t: a * scale for t, a in by_type.items()}


def estimate_savings(
    anomaly_flags: list[dict[str, Any]],
    monthly_spend: list[dict[str, Any]],
    period_total_spend: float | None = None,
) -> dict[str, Any]:
    months = _months_in_dataset(monthly_spend)
    period_spend = period_total_spend
    if period_spend is None:
        period_spend = sum(float(m.get("totalCost") or 0) for m in monthly_spend)

    by_type = cap_flag_amounts_by_spend(anomaly_flags, period_spend)

    opportunities = []
    for flag_type, period_amount in by_type.items():
        rates = RECOVERY_RATE.get(flag_type, {"low": 0.2, "high": 0.5})
        annualized = (period_amount / months) * 12
        opportunities.append(
            {
                "type": flag_type,
                "periodAmount": period_amount,
                "annualizedLow": annualized * rates["low"],
                "annualizedHigh": annualized * rates["high"],
            }
        )

    low = sum(o["annualizedLow"] for o in opportunities)
    high = sum(o["annualizedHigh"] for o in opportunities)
    annualized_cap = (period_spend / months) * 12 if period_spend > 0 else 0.0

    if annualized_cap > 0 and high > annualized_cap:
        scale = annualized_cap / high
        opportunities = [
            {
                **o,
                "annualizedLow": o["annualizedLow"] * scale,
                "annualizedHigh": o["annualizedHigh"] * scale,
            }
            for o in opportunities
        ]
        low *= scale
        high = annualized_cap

    return {
        "low": round(low, 2),
        "high": round(high, 2),
        "annualizedBasisMonths": months,
        "opportunities": sorted(opportunities, key=lambda x: -x["annualizedHigh"]),
    }
